src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseSearchValues.cs

```diff
@@ -0,0 +1,314 @@
+
+        protected abstract bool IsConstantByteOrCharArrayCreationSyntax(SemanticModel semanticModel, SyntaxNode syntax, out int length);
+
+        protected abstract bool ArrayFieldUsesAreLikelyReadOnly(SyntaxNode syntax);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffersSearchValues, out _) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlySpan1, out var readOnlySpanType))
+                {
+                    return;
+                }
+
+                var indexOfAnyMethods = GetIndexOfAnyMethods(context.Compilation);
+
+                context.RegisterOperationAction(context => AnalyzeInvocation(context, indexOfAnyMethods, readOnlySpanType), OperationKind.Invocation);
+            });
+        }
+
+        private void AnalyzeInvocation(OperationAnalysisContext context, HashSet<IMethodSymbol> indexOfAnyMethodsToDetect, INamedTypeSymbol readOnlySpanType)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+
+            if (!indexOfAnyMethodsToDetect.Contains(invocation.TargetMethod))
+            {
+                return;
+            }
+
+            Debug.Assert(invocation.Arguments.Length is 1 or 2);
+            IArgumentOperation valuesArgument = invocation.Arguments[^1];
+
+            bool isStringIndexOfAny = invocation.TargetMethod.ContainingType.SpecialType == SpecialType.System_String;
+
+            if (isStringIndexOfAny && invocation.Parent is IConditionalAccessOperation)
+            {
+                // We don't flag uses like "string?.IndexOfAny(char[])"
+                // as it's not trivial to rewrite to the span-based variant.
+                return;
+            }
+
+            if (isStringIndexOfAny
+                ? AreConstantValuesWorthReplacingForStringIndexOfAny(valuesArgument.Value)
+                : AreConstantValuesWorthReplacing(valuesArgument.Value, readOnlySpanType))
+            {
+                context.ReportDiagnostic(valuesArgument.CreateDiagnostic(Rule));
+            }
+        }
+
+        private static HashSet<IMethodSymbol> GetIndexOfAnyMethods(Compilation compilation)
+        {
+            var methods = new HashSet<IMethodSymbol>();
+
+            var stringType = compilation.GetSpecialType(SpecialType.System_String);
+
+            // string.{Last}IndexOfAny(char[])
+            // Overloads that accept 'startOffset' or 'count' are excluded as they can't be trivially converted to AsSpan.
+            foreach (var method in stringType.GetMembers().OfType<IMethodSymbol>())
```

Comment:
Consider removing the `.OfType()` invocation and instead add a type check within the if statement within the loop